### PR TITLE
Add autocompletes for case properties in case update rules and conditional alerts

### PR DESCRIPTION
**Pull request by orangejenny**
## Product Description

https://dimagi-dev.atlassian.net/browse/USH-1921

Replaces text inputs with autocompletes for the case property inputs when adding/editing automatic case update rules and conditional alerts.

These are populated based on the data dictionary. Projects without the data dictionary feature flag will still see the old text inputs. They also accept free text, in case the data dictionary is not up to date.

Screen Shot 2022-07-01 at 2 52 23 PM

## Feature Flag
Data dictionary

### Safety story
UI-only change, isolated to data dictionary projects, though there are a lot of those.

### Automated test coverage

At this UI level, none.

### QA Plan

https://dimagi-dev.atlassian.net/browse/QA-4318

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

File: corehq/apps/data_interfaces/templates/data_interfaces/partials/case_rule_case_property_input.html [hunk 1]
```diff
@@ -0,0 +1,8 @@
+{% load hq_shared_tags %}
+{% load i18n %}
+
+{% if request|toggle_enabled:"DATA_DICTIONARY" %}
```
Comment (esoergel) [inline corehq/apps/data_interfaces/templates/data_interfaces/partials/case_rule_case_property_input.html:4]: Nice, I was gonna suggest this

File: corehq/apps/hqwebapp/static/hqwebapp/js/select2_knockout_bindings.ko.js [hunk 2]
```diff
@@ -10,6 +10,28 @@ hqDefine("hqwebapp/js/select2_knockout_bindings.ko", [
...
+        var self = {};
+
+        self.init = function (element) {
+            $(element).select2({
+                width: "element",
+            });
+        };
+
+        self.update = function (element, valueAccessor, allBindings) {
+            // Do nothing, this should never get called
```
Comment (esoergel) [inline corehq/apps/hqwebapp/static/hqwebapp/js/select2_knockout_bindings.ko.js:29]: should this throw or something, or is it fine?

**Review by esoergel — APPROVED** [review]
This is great!  What's the UX if they want to type in a property that isn't in the DD list?  Does that still work well enough?

**Comment by orangejenny** [comment]
> What's the UX if they want to type in a property that isn't in the DD list?

yeah, you can still type whatever in the box

File: corehq/apps/hqwebapp/static/hqwebapp/js/select2_knockout_bindings.ko.js (hunk 2 above)
Comment (orangejenny) [reply corehq/apps/hqwebapp/static/hqwebapp/js/select2_knockout_bindings.ko.js:29]: I'd rather err on the side of the dropdown not getting updated than risk breaking the page.

Comment (orangejenny) [reply corehq/apps/hqwebapp/static/hqwebapp/js/select2_knockout_bindings.ko.js:29]: This probably shouldn't be a binding at all, I don't know what the knockout-y way is to handle "execute some code when this element is rendered."

File: corehq/apps/data_dictionary/util.py [hunk 3]
```diff
@@ -111,6 +111,10 @@ def _create_properties_for_case_types(domain, case_type_to_prop):
 
     CaseProperty.objects.bulk_create(new_case_properties)
```
Comment (orangejenny) [inline corehq/apps/data_dictionary/util.py:112]: `bulk_create` doesn't call `save`: https://docs.djangoproject.com/en/1.8/ref/models/querysets/#bulk-create

**Review by esoergel — APPROVED** [review]

**Review by esoergel — APPROVED** [review]

**Comment by orangejenny** [comment]
QA passed, re-requesting review because the last few commits have some significant refactoring.

File: corehq/apps/data_interfaces/static/data_interfaces/js/case_property_input.js [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+ *   which should be an object where keys are case type names and values are arrays
+ *   of case property names.
+ * - The calling code must call the `register` function to initialize the component
+ *   Note this must be done on DOM ready, so taht initial page data is available.
+ * - The widget itself is a custom HTML element:
+ *     <case-property-input params="
+ *       valueObservable: name,
+ *       caseTypeObservable: $root.caseType,
+ *     "></case-property-input>
+ */
```
Comment (esoergel) [inline corehq/apps/data_interfaces/static/data_interfaces/js/case_property_input.js:25]: :star2: This is great

File: corehq/apps/data_interfaces/static/data_interfaces/js/case_property_input.js [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    toggles
+) {
+    var component = {
+        viewModel: function (params) {
+            var self = this;
+
+            self.caseTypeObservable = params.caseTypeObservable;
+            self.valueObservable = params.valueObservable;
+
+            self.allCaseProperties = initialPageData.get("all_case_properties");
```
Comment (esoergel) [inline corehq/apps/data_interfaces/static/data_interfaces/js/case_property_input.js:47]: Thoughts on making this strict, to make sure implementers remember to supply it?
```suggestion
            self.allCaseProperties = initialPageData.get("all_case_properties", true);
```

**Review by esoergel — APPROVED** [review]

File: corehq/apps/data_interfaces/static/data_interfaces/js/case_property_input.js (hunk 5 above)
Comment (orangejenny) [reply corehq/apps/data_interfaces/static/data_interfaces/js/case_property_input.js:47]: I forgot about strict initial page data. But I think that similar to [this](https://github.com/dimagi/commcare-hq/pull/31854#discussion_r912212573), I'd rather the failure mode be "no autocomplete" than a js error.

**Pull request merged**